From d3da00817202a762de5bcd7cf14c38dceb30a301 Mon Sep 17 00:00:00 2001
From: Dominik Csapak <d.csapak@proxmox.com>
Date: Fri, 19 May 2017 08:53:56 +0200
Subject: [PATCH 11/11] add width and height to vncproxy call in console tab

to set the resolution we want

Signed-off-by: Dominik Csapak <d.csapak@proxmox.com>
---
 app/pve.js | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/app/pve.js b/app/pve.js
index 668835a..c10df4e 100644
--- a/app/pve.js
+++ b/app/pve.js
@@ -52,6 +52,13 @@ export default function PVEUI(UI){
 	    break;
     }
 
+    if (this.resize == 'scale' &&
+	(this.consoletype === 'lxc' || this.consoletype === 'shell')) {
+	var size = this.getFBSize();
+	params.width = size.width;
+	params.height = size.height;
+    }
+
     this.baseUrl = baseUrl;
     this.url = url;
     this.params = params;
-- 
2.11.0

