From bc515ecef9162774bff846e362b4c8315032ea2a Mon Sep 17 00:00:00 2001
From: Tim Marx <t.marx@proxmox.com>
Date: Fri, 21 Dec 2018 11:46:13 +0100
Subject: [PATCH 12/12] pass custom command to vnc

Signed-off-by: Tim Marx <t.marx@proxmox.com>
---
 app/pve.js | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/app/pve.js b/app/pve.js
index e2c37fb..9cf4b09 100644
--- a/app/pve.js
+++ b/app/pve.js
@@ -11,6 +11,7 @@ export default function PVEUI(UI){
     this.vmname = WebUtil.getQueryVar('vmname');
     this.nodename = WebUtil.getQueryVar('node');
     this.resize = WebUtil.getQueryVar('resize');
+    this.cmd = WebUtil.getQueryVar('cmd');
     this.lastFBWidth = undefined;
     this.lastFBHeight = undefined;
     this.sizeUpdateTimer = undefined;
@@ -47,6 +48,12 @@ export default function PVEUI(UI){
 	    params.upgrade = 1;
 	    title = 'System upgrade on node ' + this.nodename;
 	    break;
+	case 'cmd':
+	    url =  baseUrl + '/vncshell';
+	    params.cmd = decodeURI(this.cmd);
+	    title = 'Install Ceph on node ' + this.nodename;
+
+	    break;
 	default:
 	    throw 'implement me';
 	    break;
-- 
2.11.0

