Index: new/include/websock.js
===================================================================
--- new.orig/include/websock.js	2014-06-13 07:20:22.000000000 +0200
+++ new/include/websock.js	2014-06-18 14:12:20.000000000 +0200
@@ -282,7 +282,7 @@
     // TODO: this sucks, the property should exist on the prototype
     // but it does not.
     try {
-        if (bt && ('binaryType' in (new WebSocket("ws://localhost:17523")))) {
+        if (bt && ('binaryType' in (new WebSocket("wss://localhost:17523")))) {
             Util.Info("Detected binaryType support in WebSockets");
             wsbt = true;
         }
