--- a/pveui.js
+++ b/pveui.js
@@ -1156,6 +1156,7 @@
                 $D('toggleAltButton').style.display = "inline";
                 $D('sendTabButton').style.display = "inline";
                 $D('sendEscButton').style.display = "inline";
+		$D('sendCtrlAltDelButton').style.display = "inline";
                 $D('showExtraKeysButton').className = "noVNC_status_button_selected";
                 UI.extraKeysVisible = true;
             } else if(UI.extraKeysVisible === true) {
@@ -1163,6 +1164,7 @@
                 $D('toggleAltButton').style.display = "";
                 $D('sendTabButton').style.display = "";
                 $D('sendEscButton').style.display = "";
+		$D('sendCtrlAltDelButton').style.display = "";
                 $D('showExtraKeysButton').className = "noVNC_status_button";
                 UI.extraKeysVisible = false;
             }
